lib/services/cordova-migration-service.ts

```diff
@@ -134,6 +143,153 @@ export class CordovaMigrationService implements ICordovaMigrationService {
+				this.$errors.failWithoutHelp("The selected version %s is not supported. Supported versions are %s", newVersion, validWPSdks.join(", "));
+			}
+
+			this.$logger.info("Migrating to WPSdk version %s", newVersion);
+			if(helpers.versionCompare(newVersion, "8.0") > 0) {
+				// at least Cordova 3.7 is required
+				if(helpers.versionCompare(this.$project.projectData.FrameworkVersion, "3.7.0") < 0) {
+					let cordovaVersions = this.getSupportedFrameworks().wait();
+
+					// Find last framework which is not experimental.
+					let selectedFramework = _.findLast(cordovaVersions, cv => cv.DisplayName.indexOf(this.$projectConstants.EXPERIMENTAL_TAG) === -1);
+					if(helpers.versionCompare(selectedFramework.Version, "3.7.0") < 0) {
+						// if latest stable framework version is below 3.7.0, find last 'Experimental'.
+						selectedFramework = _.findLast(cordovaVersions, cv => cv.DisplayName.indexOf(this.$projectConstants.EXPERIMENTAL_TAG) !== -1 && helpers.versionCompare("3.7.0", cv.Version) <= 0);
+					}
+
+					let shouldUpdateFramework = this.$prompter.confirm(`You cannot build with the Windows Phone ${newVersion} SDK with the currently selected target version of Apache Cordova. Do you want to switch to ${selectedFramework.DisplayName}?`).wait()
+					if(shouldUpdateFramework) {
+						this.onFrameworkVersionChanging(selectedFramework.Version).wait();
+						this.$project.projectData.FrameworkVersion = selectedFramework.Version;
+					} else {
+						this.$errors.failWithoutHelp("Unable to set Windows Phone %s as the target SDK. Migrate to Apache Cordova 3.7.0 or later and try again.", newVersion);
+					}
+				}
+			}
+		}).future<void>()();
+	}
+
+	public onFrameworkVersionChanging(newVersion: string): IFuture<void> {
+		return ((): void => {
+			if(newVersion === this.$project.projectData.FrameworkVersion) {
+				return;
+			}
+
+			this.$project.ensureAllPlatformAssets().wait();
+
+			if(this.$project.projectData.WPSdk && helpers.versionCompare(this.$project.projectData.WPSdk, "8.0") > 0 && helpers.versionCompare(newVersion, "3.7.0") < 0) {
+				let shouldUpdateWPSdk = this.$prompter.confirm(`You cannot build with the Windows Phone ${this.$project.projectData.WPSdk} SDK with the currently selected target version of Apache Cordova. Do you want to switch to Windows Phone 8.0 SDK?`).wait();
+				if(shouldUpdateWPSdk) {
+					this.onWPSdkVersionChanging("8.0").wait();
+					this.$project.projectData.WPSdk = "8.0";
+				} else {
+					this.$errors.failWithoutHelp("Unable to set %s as the target Apache Cordova version. Set the target Windows Phone SDK to 8.0 and try again.", newVersion);
+				}
+			}
+
+			let versionDisplayName = this.getDisplayNameForVersion(newVersion).wait();
+			this.$logger.info("Migrating to Cordova version %s", versionDisplayName);
+			let oldVersion = this.$project.projectData.FrameworkVersion;
+			
+			let allMarketplacePlugins: string[] = [];
+			_.each(this.$project.configurations, configuration => {
```

Comment:
allMarketplacePlugins is actually only used once - to filter out invalid plugins. Can this operation be chained in too?